### PR TITLE
Product Channels Fixes

### PR DESCRIPTION
**Pull request by TomRadford**

### CODE REVIEW

File: src/models/channels/system-channel.ts [hunk 1]
```diff
@@ -1,45 +1,12 @@
...
-    price_tier,
-    qty_availability,
-    sync_token
-  }: Partial<SystemChannel> = {}) {
-    this.id = id;
-    this.created = created;
-    this.description = description;
-    this.modified = modified;
-    this.client_id = client_id;
-    this.active = active ?? 0;
```
Comment (TomRadford) [inline src/models/channels/system-channel.ts:39]: this one of the sources of that funky boolean being 0/1 issue

File: src/components/Product/ProductChannels.svelte [hunk 2]
```diff
@@ -1,97 +1,71 @@
...
-      enabled: !!productSourceId
-    }
+  $: productChannelData = createProductChannelsQuery(
+    parseInt($page.params.id),
+    !!$product.workingCopy?.source_id
   );
 
-  const fetchVariantChannels = async (variantId, channel_id) => {
+  //Start vm creator block
+  let productVariantChannels: (
```
Comment (TomRadford) [inline src/components/Product/ProductChannels.svelte:58]: we use this var to store only required variant channels. We push to it from processChannel. 
It's a little hard to follow but seems to work well with the tweaks I made. Just need to reset it whenever re-running the VM creator query.

File: src/components/Header/Tabs.svelte [hunk 3]
```diff
@@ -7,7 +7,7 @@
 
   export let tabs: { name: string; key: string }[];
   export let baseUrl: string;
-  export let allowNavigation: () => boolean;
+  export let allowNavigation = () => true;
```
Comment (Ayano2000) [inline src/components/Header/Tabs.svelte:10]: @TomRadford is going to remove the `allowNavigation` in favor of `beforeNavigation`

**Review by Ayano2000 — CHANGES_REQUESTED** [review]
See: https://github.com/stock2shop/console_v2/pull/473/files#r1498814715

File: src/components/Product/ProductChannels.svelte [hunk 4]
```diff
@@ -176,45 +190,132 @@
       });
     });
 
-    return changedVariants;
+    const updateVariantChannelPromises = changedVariants.map(async (cv) => {
+      const post: SystemVariantChannelPost = {
+        system_variant_channel: {
+          channel_id: parseInt(cv.channel_id),
+          channel_variant_code: cv.channel_variant_code
```
Comment (matthampton-stock2shop) [inline src/components/Product/ProductChannels.svelte:197]: I had some odd behaviour when the text fields where blank (i.e. cv.channel_variant_code === '') - where it would create a row in the db with a blank value - bit then it would not delete them.

I think we should do something about this

File: src/components/Header/Tabs.svelte (hunk 3 above)
Comment (TomRadford) [reply src/components/Header/Tabs.svelte]: 927ec7a0eb949ff70905d786ddf4a806f0318a74

File: src/queries/channelSources.query.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import { createQuery } from '@tanstack/svelte-query';
+
+const channelSourcesRepo = new ChannelSourcesRepo();
+
+// Query Keys Factory
+export const channelSourcesQueryKeys = {
+  id: (id: number) => ['channel-sources', id]
+};
+
+export const createChannelSourceQuery = (id: number, enabled = true) =>
```
[the remark this answers is not in the document]
Comment (TomRadford) [reply src/queries/channelSources.query.ts:11]: Going to be quite complex to get this to work because of all the data mapping going on. I think we can do this one when the new API is in place.

File: src/routes/(app)/customer/[id]/+layout.svelte [hunk 6]
```diff
@@ -102,15 +102,22 @@
 
     $isSaving = true;
 
+    let continueAfterOnSave = true;
+
     if ($onSaveChanges) {
       try {
-        await $onSaveChanges();
+        continueAfterOnSave = await $onSaveChanges();
```
[the remark this answers is not in the document]
Comment (TomRadford) [reply src/routes/(app)/customer/[id]/+layout.svelte]: hmm the purpose of onSaveChanges returning false it to tell the save function to not run the rest of the code. A failure is indicated with a throw

File: src/routes/(app)/product/[id]/+layout.svelte [hunk 7]
```diff
@@ -64,12 +70,12 @@
...
-    if ($product.changesMade) {
-      return confirm('Are you sure you want to navigate away? Changes will be lost.');
-    }
-    return true;
-  }
+  const allChangesMade = writable(false);
+  $: $allChangesMade = $product.changesMade || $changesMade;
+  safeNavigationCheck(allChangesMade, () => {
+    $changesMade = false;
+    $product.workingCopy = cloneDeep($product.serverVersion);
```
[the remark this answers is not in the document]
Comment (TomRadford) [reply src/routes/(app)/product/[id]/+layout.svelte:77]: 927ec7a0eb949ff70905d786ddf4a806f0318a74

File: src/routes/(app)/customer/[id]/+layout.svelte (hunk 6 above)
[the remark this answers is not in the document]
Comment (TomRadford) [reply src/routes/(app)/customer/[id]/+layout.svelte]: 246f69c567e18c8bd4d1d61ae8b61f8f2cf49f11

File: src/routes/(app)/product/[id]/+layout.svelte [hunk 8]
```diff
@@ -64,12 +70,12 @@
     { name: 'Dynamic Pricing', key: 'dynamic-pricing' }
   ] as const;
 
-  function allowNavigation() {
-    if ($product.changesMade) {
-      return confirm('Are you sure you want to navigate away? Changes will be lost.');
-    }
-    return true;
-  }
+  const allChangesMade = writable(false);
```
[the remark this answers is not in the document]
Comment (TomRadford) [reply src/routes/(app)/product/[id]/+layout.svelte:73]: 1d74e3b96e70d1b4900ef4f72329cdaaad7e283f

File: src/routes/(app)/product/[id]/+layout.svelte [hunk 9]
```diff
@@ -251,7 +264,7 @@
         key: 's',
         modifier: ['ctrl', 'meta'],
         preventDefault: true,
-        enabled: $product.changesMade,
+        enabled: $allChangesMade || !$isSaving,
```
[the remark this answers is not in the document]
Comment (TomRadford) [reply src/routes/(app)/product/[id]/+layout.svelte]: yes, resolved

File: src/components/Product/ProductChannels.svelte (hunk 4 above)
Comment (TomRadford) [reply src/components/Product/ProductChannels.svelte:197]: b68b4732964116bfafb8a55f734b26598348e79c

**Review by Ayano2000 — APPROVED** [review]
LGTM
